### PR TITLE
Add caml_plat_lock_non_blocking & compatiblity between sync.h and platform.h

### PR DESCRIPTION
**Pull request by gadmm**
For the purposes of #12410, the runtime needs a version of `caml_plat_lock` that does not block. (Or equivalently, a version of `caml_ml_mutex_lock` that does not raise exceptions.)

In addition, the mutex in question comes from OCaml, but it is important to lock it from C. For the same purposes, we need to manipulate from C a condition variable coming from OCaml, again without raising exceptions.

In order to add the required functionality without making a mess, the plan is the following:
- Make mutexes and condvars from sync.h and platform.h explicitly compatible. (second commit)
- This requires changing a bit condvars from platform.h, but the reason for the discrepancy with sync.h is not convincing (cf. commit log). We gain in simplicity by aligning platform.h on sync.h with respect to condvars (first commit).
- Add the required function `caml_plat_lock_non_blocking`, document it, and use it at one place where this functionality was already required. (third commit)

An ulterior PR will introduce more usages of `caml_plat_lock_non_blocking` by auditing all current uses of `caml_plat_lock`.

No change entry needed (credit to reviewers can be given inside the Changes for #12410).

cc @kayceesrk

### CODE REVIEW

**Comment by gadmm** [comment]
cc @dra27 as well, in case this interferes with plans to bring OCaml 5 to non-POSIX platforms (e.g. Win32)

**Comment by gadmm** [comment]
I do not believe that this PR introduces any observable change. Is `extra (debug)` known to be flaky on current trunk?

**Comment by gadmm** [comment]
The added commit fixes `caml_plat_assert_all_locks_unlocked`. The choice of static storage duration meant that is was always verified, that is until I tried using `DEBUG_LOCK` in this compilation unit.

This should fix the `extra (debug)` CI.

File: runtime/caml/platform.h
Comment (ghost) [inline runtime/caml/platform.h]:
```suggestion
Caml_inline void caml_plat_lock_non_blocking(caml_plat_mutex*);
```
for consistency with the other prototypes in this file.

**Review by ghost — APPROVED** [review]
LGTM

**Comment by gadmm** [comment]
Rebased & cleaned-up. A documentation comment has been clarified. I'd be happy to see some progress on this PR.

File: runtime/caml/platform.h
Comment (eutro) [inline runtime/caml/platform.h]: It might make more sense, I think, to write this in terms of `caml_enter_blocking_section` and `caml_leave_blocking_section`, both since it describes what it does more precisely ("I can block within this section and it's OK"), and because the domain lock is sort of an implementation detail, which is less documented ("what is the domain lock? what does it protect?").

My understanding is that the issue is blocking without any C thread to handle possible interrupts (neither the main nor backup thread). It seems the domain lock must be held for handling interrupts (I suppose to ensure that the main thread cannot enter re-enter the mutator while the backup thread is handling interrupts, and that the backup thread doesn't handle interrupts before the mutator has been left). However it is not specifically the holding of the lock that prevents the backup thread from handling interrupts, since the backup thread must also be properly be in the `BT_IN_BLOCKING_SECTION` state.

**Review by eutro — COMMENTED** [review]
I think `caml_plat_lock_non_blocking` is a somewhat misleading name. It does block: it blocks the C thread it's on, it just does this in a blocking section. Is there specific reasoning for using "non_blocking"? Something like `caml_plat_lock_from_mutator`, perhaps, seems more intuitive to me.

File: runtime/caml/platform.h
Comment (gadmm) [reply runtime/caml/platform.h]: You correctly point out a subtlety about domain locks. In OCaml 4, the runtime lock was essentially a mutex, but in OCaml 5 the domain lock is an abstraction that also deals with waking up the backup thread.
That being said, a clarification about this would be better located in the manual about the domain lock (https://ocaml.org/manual/5.1/intfc.html#ss:parallel-execution-long-running-c-code).
I think this documentation is fine here because *domain lock*, `caml_release/acquire_runtime_system`, and `caml_enter_blocking_section/caml_enter_blocking_section` are described as corresponding to the same notion in the manual.

**Comment by gadmm** [comment]
I am open to alternative names. But:
- The non-non_blocking function are documented as `/* blocking */` in the sense that they might need to appear in a blocking section. In that sense the opposite property is non-blocking, which can be understood as not blocking the domain.
- All the new functions and documentation are internal, so naming is not a big deal as long as it does not cause confusion for OCaml maintainers.

Let me know what you think. By the way, thanks for looking at this. I saw your other PR about OS-based synchronisation and I was wondering what you thought about this issue.

File: runtime/caml/platform.h
Comment (gasche) [inline runtime/caml/platform.h:118]: It the loss of the documentation comment here intentional?

File: runtime/caml/platform.h
Comment (gasche) [inline runtime/caml/platform.h:136]: My impression is that `caml_plat_lock` is a somewhat dangerous operation (it is unsafe to run a STW inside the critical section, due to the deadlocking bugs similar to #1287), and `caml_plat_lock_non_blocking` should be the default. Maybe the first one should be called `caml_plat_lock_blocking` or `caml_plat_block_critical`?

Taking a step back, the two functions serve fairly different purposes:
- `caml_plat_lock` is for low-level exclusion / short critical sections inside the runtime code
- `caml_plat_non_blocking` can be used as a C counterpart of `Mutex.lock` in the OCaml code, allowing to run arbitrary OCaml (and C) code

I am not sure that the naming reflects this. Maybe this could be `caml_plat_lock_unsafe` and `caml_mutex_lock`, or something else.

File: runtime/caml/platform.h
Comment (gasche) [inline runtime/caml/platform.h:137]: Currently you use `caml_plat_unlock` in both cases, whether `caml_plat_lock` or `caml_plat_lock_blocking` was used. I see how that works, but I am a bit nervous about this being inflexible in the future -- say if we want to change the implementations in different ways, or if we want different debug/instrumentation logic for both sorts of critical section. Could we have two different `unlock` functions?

**Review by gasche — COMMENTED** [review]
I understand the changes here and I agree that they go in a good direction -- having this available would have made our life much easier when we fixed #12897, a deadlock bug in runtime_events caused by `caml_plat_lock` .

I would be in favor of a larger separation between the two locking functions/modes (the "blocking" and "non-blocking" variants), possibly with different names and also with a different unlocking function.

Overall my impression is that this is okay, but maybe @gadmm could do a round of API naming/thinking before we converge to merge the PR.

File: runtime/caml/platform.h
Comment (gadmm) [reply runtime/caml/platform.h:118]: Yes, caml_plat_timed_wait has not been defined here for a while.

File: runtime/caml/platform.h
Comment (gadmm) [reply runtime/caml/platform.h:136]: I agree with your step back. Regarding naming, it is important to me that it is clear to the reader that `caml_plat_lock_non_blocking` might execute code concurrently on the same domain, which is crucial for reasoning about correctness of the runtime implementation.

What about `caml_plat_lock_blocking` and `caml_plat_lock_non_blocking`?

File: runtime/caml/platform.h
Comment (gadmm) [reply runtime/caml/platform.h:137]: It is fundamentally the same unlocking function—even `Mutex.unlock` does not release the domain lock. Also we are doing an internal API, so there is no need to worry about future-proofing.

File: runtime/caml/platform.h
Comment (gadmm) [reply runtime/caml/platform.h:136]: > What about caml_plat_lock_blocking and caml_plat_lock_non_blocking?

ping

Comment (gasche) [reply runtime/caml/platform.h:136]: I agree with your proposed names, apologies for being unresponsive earlier.

**Comment by gadmm** [comment]
I rebased and added a commit to rename `caml_plat_lock` into `caml_plat_lock_blocking` as discussed. Note that auditing whether these calls should be replaced by calls to `caml_plat_lock_non_blocking` is still scheduled for a later PR.

File: runtime/caml/platform.h
Comment (gasche) [inline runtime/caml/platform.h]: This should be `caml_plat_lock_blocking` now.

File: runtime/caml/platform.h
Comment (gasche) [reply runtime/caml/platform.h]: Personally I find the terminology "{enter,leave} a blocking section" confusing, because "blocking section" may be interpreted as meaning two opposite things, "a section that blocks the runtime", or "a section that contains a blocking operation while making sure that the runtime is not blocked". The notion of taking/releasing the {runtime, domain} lock are slightly more low-level but also clearer in what they mean, so I find them easier to reason about.

My mental model:
- mutator threads must "own a domain" to use runtime services
- taking ownership is realized by taking the domain lock
- the owner of the domain is also responsible for handling STW interrupts, so it must ensure that it remains available to do so

In programs that run several threads on a single domain, releasing the domain lock lets other threads run and handle interrupts. In any case, the runtime ensures that there is always at least one other thread that is available (the backup thread).

**Review by gasche — APPROVED** [review]

**Comment by gasche** [comment]
(I think you could squash your last commit in the previous one; otherwise the history is clean.)

**Comment by gadmm** [comment]
I think it's fine.

**Comment by gasche** [comment]
Note: is it intentional to not have a Changes entry?

(If you add one, please do squash the last commit.)

**Comment by gadmm** [comment]
Yes, I proposed to credit reviewers as part of #12410.

**Pull request merged**
